### ISSUE TITLE
compilation issue and possible output issue with different version

### ISSUE THREAD

**Issue by mpr121** [jlmelville/uwot#75]
failed to install 0.1.8 in R 4.0.3 in Rstudio(compilation error).  Different version have different umap output. 
details:
 I noticed that different version of uwot could produce different results.  e.g. the umap output of 0.1.10(my current version) was different from that of 0.1.8. To reproduce figures made by earlier uwot version, I tried to install the corresponding one, but it failed.  Rtools40 in R4.0.3 was installed and worked correctly in compiling other packages. If the installation issue can't be fixed, then can you advice how to make the outputs of uwot consistent across different version?
thanks,

**Comment by jlmelville** [comment]
@mpr121 please provide the error output you are getting trying to build 0.1.8, in case it's something that can be fixed.

Knowing the commands you ran would help as would whether it was on the same machine/operating system and versions of R. You won't get reproducible results if the `n_sgd_threads` is greater than 1 or if you didn't `set.seed` to a fixed value. I also doubt whether results would be reproducible across different architectures. I also don't know what the expectation would be going from R 3.x to R 4.x. 

You do of course know your data better than anyone else, but more broadly, I would be exceptionally nervous if the variability between different runs of `umap` was causing you to draw different conclusions from your data. That would be a strong indication to me that any pattern in the data was an artifact. If the relationship between the data you originally saw doesn't show up often when you repeat the calculation, there is a strong possibility there is no real relationship.

In short, there's no reason to believe the result you got from version 0.1.8 is more "real" than what you are getting now. If exact reproducibility is required (again I think this is a very bad idea), then "freezing" an entire Linux installation (including distribution version, installed software, R version and other packages) via a container is your best solution.

**Comment by mpr121** [comment]
Thank you very much for your quick reply, James!  

the issue I encountered is that the output from different versions of uwot after running a command RunUMAP (a function from Seurat package) appeared to be different.  The same version of uwot method should produce the same UMAP figure.  
 
Regarding installation issue, here is the command line I used to install older versions. 
devtools::install_version("uwot", version = "0.1.8") 

The errors at the last few lines: 
make: *** [C:/PROGRA~1/R/R-40~1.3/etc/i386/Makeconf:229: nn_parallel.o] Error 1
ERROR: compilation failed for package 'uwot'
* removing 'C:/Users/xx/Documents/R/win-library/4.0/uwot'
Error: Failed to install 'unknown package' from URL:
  (converted from warning) installation of package 'C:/Users/xx/AppData/Local/Temp/RtmpEpUFYQ/remotes11606b4563be/uwot' had non-zero exit status

I compared the seurat object in the .Rdata file provided by the author of a published paper with the object after I ran the RunUMAP command using the same code and the same dataset. I found the only different part of the s4 seurat objects is the following:

from version 0.1.8(result from the published paper):  
.. ..$   umap:Formal class 'DimReduc' [package "Seurat"] with 9 slots |  
.. .. .. ..@ cell.embeddings           : num [1:1385, 1:2] -6.09 -6.79   -7.64 4.91 8.66 ...
.. .. .. .. ..- attr(*,   "scaled:center")= num [1:2] 4.1 -3.2 

version 0.1.10 (the result after I ran the code):
.. ..$ umap:Formal class 'DimReduc'   [package "Seurat"] with 9 slots
.. .. .. ..@ cell.embeddings           : num [1:1385, 1:2] 12.5 11.6 11.1   -10.6 -11 ...
.. .. .. .. ..- attr(*,   "scaled:center")= num [1:2] -1.17 -3.14

Since the RunUMAP uses 'uwot' package as the method by default and I found that my uwot version was different from the author's, and the only different part between the seurat objects was the UMAP output, I guess the 'uwot' version could be the most likely cause. Therefore, I tried to install older version of 'uwot' to test the possibility. Maybe I am wrong. I was surprised to see how the same code and original dataset could produce different figure pattern, though still similar to some extent(new figure has shuffled clusters and patterns). 

Thank you for your helps!

**Comment by LTLA** [comment]
Looks like you're using 32-bit Windows, based on the i386 in your path. If so, another possible explanation for the discrepancy is the different numerical precision used for the internal math. This can introduce very small differences at each step that snowball into much larger differences in iterative algorithms like t-SNE (and possibly also UMAP, though I have not tested this). Differences in precision on 32-bit is a common source of headaches for **scran**'s unit tests on Bioconductor's build system.

If this is the cause, then there's little you can do beyond getting access to a 64-bit environment, preferably running Linux.

**Comment by mpr121** [comment]
the system is win10, 64-bit operating system, x64-based processor. Rtools is rtools40-x86_64.exe for 64 bit system.

**Comment by jlmelville** [comment]
@mpr121 the error output cuts off before showing all the compilation error, but the problem there is that you will also need an older version of the RcppAnnoy package, as I had to make breaking changes to some C++ internals in uwot to support changes in RcppAnnoy's internals, which in turn had been slightly modified to keep up with changes to the underlying Annoy library. I don't suggest going down that path unless you use the container approach.

Currently, I can't commit to keeping UMAP output reproducible between version changes. I don't know if that is even feasible due to the different ways the parts of the method works (e.g. random number generation, iterative precision differences accumulating errors as @LTLA mentions, changes to nearest neighbor search libraries, differences in PCA or spectral initialization).

If exact numerical reproducibility is critical for your needs I can't recommend using uwot, or maybe any UMAP implementation. I don't even know if the Python version is reproducible between versions or changes in dependencies such as numba. I'd consider t-SNE as an alternative as long as you initialize it suitably. It should be less susceptible (but not entirely immune) to these issues due to not using a stochastic optimization.

**Comment by mpr121** [comment]
It is not a big issue. I agree not to expect exact numerical outputs considering there are certain stochastic processes.   The compilation issue was solved with the a correct version of RcppAnnoy package. Glad to know that.  Thank you. Cool!
